sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -71,6 +71,11 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
 
   val id: Int = SparkPlan.newPlanId()
 
+  /**
+   * Return true if this stage of the plan supports row-based execution.
```

Comment:
> supportsColumnar: this plan can support columnar output, alongside the default row-based output which every plan supports.

Hmm, seems not exactly? Not every plan supports row-based output.

`prefersColumnar` seems redundant? As I see, we usually prefer columnar output already.

Reply:
> Not every plan supports row-based output.

Hmm any example? every physical plan node has to implement `doExecute` which outputs rows, while `doExecuteColumnar` throws exception by default.

> As I see, we usually prefer columnar output already.

I'm not sure about this part. To my understanding, at the moment it appears we prefer columnar output because 1) vectorized readers for OPC/Parquet yield much better performance so we always want to use that over the default row-based impls, and 2) `supportsColumnar` defaults to false as most operators don't support columnar execution yet, so we'll do the columnar-row conversion and switch back to whole-stage codegen.

However this may not hold true if we add columnar support for more operators like filter/project etc in future. Do we want to prefer columnar execution over the whole-stage codegen approach? I'm not sure yet and maybe some evaluation is required. `prefersColumnar` could give us a knob to control this.

Reply:
> Hmm any example? every physical plan node has to implement doExecute which outputs rows, while doExecuteColumnar throws exception by default.

I think there is no guarantee that a physical node must implement a working `doExecute`. For a columnar node, it can just throw exception saying it is not implemented (like the default `doExecuteColumnar`) if it is not designed to be executed under row-based execution.

I also don't see a need to have implement both (working) row-based and columnar execution for a node in general. But in Spark, because we don't actually have official columnar execution nodes, so maybe I cannot get an example from Spark itself. Hopefully I convey the idea clearly.

> prefersColumnar: this plan prefers to output columnar batches even if it is not explicitly requested (e.g., outputsColumnar is false).

BTW, `outputsColumnar` is not a preference option I think (at least for its usage now in the rule). It actually means that the output should be in columnar or not. Once `outputsColumnar` is false, the plan should output row-based output and it is why we add `ColumnarToRowExec` for the case.

Yea, the preference I mentioned is pretty limited so far. I agree that we maybe need to have a preference rule (or something) in the future. As we don't have real built-in columnar operators in Spark, so currently the situation seems that some columnar extensions/libraries replace row-based operators with columnar operators during planning. I'm not sure if we can estimate which one is preferred during planning.

Reply:
BTW, IMHO, if we add columnar support for more operators in the future, I guess it already implicitly indicates we "prefer" it over current execution (whole-stage codegen or interpreted one)? Just like whole-stage codegen, seems we simply prefer it once we verify it having better performance generally. This is similar to the 3rd party extensions/libraries situation, I think.

Reply:
I see, makes sense. I was referring to nodes in Spark itself but yea an extension could only implement doExecuteColumnar.

Although I'm still slightly in favor of prefersColumnar, but it's only a minor personal preference. Overall it looks OK.

Reply:
For now I guess columnar route seems to be considered as superior, otherwise there should be calculation for the cost of plan between row vs columnar.

This is just to cover the case that downstream doesn't support columnar but upstream can support both row and columnar, and has performance of producing output as `columnar output > row output > columnar output + columnar-to-row conversion`, so upstream wants to produce directly whatever downstream wants without conversion.

If upstream can produce columnar output faster enough to cover the overhead of columnar-to-row conversion, then it could just tactically say "it only cover columnar output" and Spark will take the conversion.